lib/ansible/modules/clustering/k8s/k8s_auth.py

```diff
@@ -0,0 +1,335 @@
+
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+
+module: k8s_auth
+
+short_description: Authenticate to Kubernetes clusters which require an explicit login step
+
+version_added: "2.8"
+
+author: KubeVirt Team (@kubevirt)
+
+description:
+  - "This module handles authenticating to Kubernetes clusters requiring I(explicit) authentication procedures,
+    meaning ones where a client logs in (obtains an authentication token), performs API operations using said
+    token and then logs out (revokes the token). An example of a Kubernetes distribution requiring this module
+    is OpenShift."
+  - "On the other hand a popular configuration for username+password authentication is one utilizing HTTP Basic
+    Auth, which does not involve any additional login/logout steps (instead login credentials can be attached
+    to each and every API call performed) and as such is handled directly by the C(k8s) module (and other
+    resource–specific modules) by utilizing the C(host), C(username) and C(password) parameters. Please
+    consult your preferred module's documentation for more details."
+
+options:
+  state:
+    description:
+    - If set to I(present) connect to the API server using the URL specified in C(host) and attempt to log in.
+    - If set to I(absent) attempt to log out by revoking the authentication token specified in C(api_key).
+    default: present
+    choices:
+    - present
+    - absent
+  host:
+    description:
+    - Provide a URL for accessing the API server.
+    required: true
+  username:
+    description:
+    - Provide a username for authenticating with the API server.
+  password:
+    description:
+    - Provide a password for authenticating with the API server.
+  ssl_ca_cert:
+    description:
+    - "Path to a CA certificate file used to verify connection to the API server. The full certificate chain
+      must be provided to avoid certificate validation errors."
+  verify_ssl:
+    description:
+    - "Whether or not to verify the API server's SSL certificates."
+    type: bool
+    default: true
+  api_key:
+    description:
+    - When C(state) is set to I(absent), this specifies the token to revoke.
+
+requirements:
+  - python >= 2.7
+  - urllib3
+  - requests
+  - requests-oauthlib
+'''
+
+EXAMPLES = '''
+- block:
+  # It's good practice to store login credentials in a secure vault and not
+  # directly in playbooks.
+  - include_vars: k8s_passwords.yml
+
+  - name: Log in (obtain access token)
+    k8s_auth:
+      host: https://k8s.example.com/
+      ssl_ca_cert: ca.pem
+      username: admin
+      password: "{{ k8s_admin_password }}"
+    register: k8s_auth_results
+
+  - name: Preserve auth info as both a fact and a yaml anchor for easy access later
+    # Both the fact and the anchor are called 'k8s_auth_params'
+    set_fact:
```

Comment:
This brings unnecessary complexity to user, why not make it simpler, and set such fact inside the module?

Reply:
Are you suggesting using `ansible_facts` for that (which seems like a bit of a hack?). I'm not aware of another way of transferring state from one module to other modules.

Even if using `ansible_facts` (and I don't think we should) we still need to document for people with `INJECT_FACTS_AS_VARS` set to False

Reply:
I'm pretty sure Ondra's suggesting returning to my initial code with the module accepting an `k8s_auth` dict as a parameter (as well as returning one) and having all of the k8s modules accept that.

Summary of the discussion around this is – the proper way to do this kind of thing going forward will be with with [module defaults for groups](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/user_guide/playbooks_module_defaults.rst#module-defaults-groups), which became available since 2.7. So the yaml anchor usage in the EXAMPLE is basically just a stop–gap for how this will end up being used properly in the future.

So while it made sense to implement a dedicated auth dict parameter for groups of modules pre–2.7, it really would not be the proper implementation nowadays.

Reply:
Yes, I meant using `ansible_facts`, but now I realized you can't create anchor from that I guess.
Yeah the module defaults solve this quite nice, actually, so forget my comment.